### PR TITLE
New `LeanIMT` incremental Merkle tree

### LINKED ISSUE #67: Add new `LeanIMT` incremental Merkle tree

### PR DESCRIPTION
**Pull request by cedoor**
## Description

This PR moves the old `incremental-merkle-tree` library to the new `imt` package, renaming classes and files, and adds a new incremental Merkle tree library (i.e. `LeanIMT`), an optimized binary incremental Merkle tree.

The libraries' code has been documented and tests have been reorganized to meet high-quality requirements.

## Related Issue

Closes #67 

## Does this introduce a breaking change?

-   [ ] Yes
-   [x] No

## TODO

- [x] Add more comments to the old `IMT` implementation
- [x] Rename `incremental-merkle-tree.sol` to `imt.sol`
- [x] Refactor old Solidity implementations' tests
- [x] Add `LeanIMT` Solidity implementation
- [x] Add documentation to Solidity implementations' code
- [x] Add warnings about new lib names

### CODE REVIEW

File: packages/imt.sol/contracts/package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
-    "name": "@zk-kit/incremental-merkle-tree.sol",
-    "version": "1.7.0",
-    "description": "Incremental Merkle tree Solidity libraries.",
+    "name": "@zk-kit/imt.sol",
+    "version": "2.0.0-beta",
```
Comment (0xjei) [inline packages/imt.sol/contracts/package.json:3]: Do we wanna keep the beta right now?

Comment (0xjei) [inline packages/imt.sol/contracts/package.json:3]: Do we wanna keep the beta right now?

File: packages/imt.sol/package.json [hunk 2]
```diff
@@ -22,8 +21,7 @@
         "@typechain/hardhat": "^6.1.2",
         "@types/chai": "^4.2.0",
         "@types/mocha": ">=9.1.0",
-        "@zk-kit/incremental-merkle-tree": "^0.4.1",
-        "circomlibjs": "^0.0.8",
+        "@zk-kit/imt": "workspace:packages/imt",
```
Comment (0xjei) [inline packages/imt.sol/package.json:24]: Can you verify that this import is correct?

File: packages/imt.sol/hardhat.config.ts [hunk 3]
```diff
@@ -3,9 +3,7 @@ import { config as dotenvConfig } from "dotenv"
 import { HardhatUserConfig } from "hardhat/config"
 import { resolve } from "path"
 import { config } from "./package.json"
-import "./tasks/deploy-ibt-test"
-import "./tasks/deploy-iqt-test"
-import "./tasks/deploy-lmt-test"
+import "./tasks/deploy-imt-test"
```
Comment (0xjei) [inline packages/imt.sol/hardhat.config.ts:6]: Very neat +1

File: packages/imt/README.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+tree.update(1, 2) // tree1.leaves -> [1, 2]
+
+tree.delete(1) // tree1.leaves -> [1, 0]
+
+const proof = tree.createProof(1)
+
+tree.verifyProof(proof) // true
+
+// You can also initialize a tree with a list of leaves.
+new IMT(poseidon, depth, zeroValue, arity, [1, 2, 3])
```
Comment (0xjei) [inline packages/imt/README.md:115]: nit: rename `poseidon` with `poseidon2`

File: packages/imt/README.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+tree.depth // 1
+tree.size // 2
+tree.leaves // [1n, 3n]
+
+tree.indexOf(3n) // 1
+tree.has(4n) // false
+
+tree.update(1, 2n) // tree1.leaves -> [1n, 2n]
+
+const proof = tree.createProof(1)
```
Comment (0xjei) [inline packages/imt/README.md:141]: nit: rename `createProof` with `generateProof`

File: packages/imt/README.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+new IMT(poseidon, depth, zeroValue, arity, [1, 2, 3])
+```
+
+### LeanIMT ([doc](https://zkkit.pse.dev/classes/_zk_kit_lean-imt.LeanIMT.html))
+
+```typescript
+import { LeanIMT } from "@zk-kit/imt"
+import { poseidon2 } from "poseidon-lite"
+
+const hash = (a, b) => poseidon2([a, b])
```
Comment (0xjei) [inline packages/imt/README.md:124]: nit: missing types for `a` and `b`?

**Review by 0xjei — CHANGES_REQUESTED** [review]
Outstanding job! I truly commend your approach in modularizing and generalizing the code to the fullest extent. The organization is both meticulous and efficient, to be honest :) I've noticed a few minor oversights in the sample READMEs, but overall, fantastic work! :star_struck:

File: packages/imt.sol/contracts/package.json (hunk 1 above)
Comment (cedoor) [reply packages/imt.sol/contracts/package.json:3]: Hey, yes! Until we test it with Semaphore v4 and have a audit 👍🏽

File: packages/imt.sol/package.json (hunk 2 above)
Comment (cedoor) [reply packages/imt.sol/package.json:24]: Yes it's correct!

File: packages/imt/README.md (hunk 6 above)
Comment (cedoor) [reply packages/imt/README.md:124]: I think this is fine without types

**Pull request merged**